### PR TITLE
Make migration context public

### LINKED ISSUE #36544: Need a public method for migrating database in tests
For gems that use database migrations, many of them were previously using this code in a `test_helper.rb` file:

```
ActiveRecord::Migrator.migrate File.expand_path("../dummy/db/migrate/", __FILE__)
```

This method was removed and now people are using:

```
 ActiveRecord::MigrationContext.new('spec/dummy/db/migrate').migrate
```

For example https://github.com/influitive/apartment/pull/523/files, however this is also not a public interface and has changed again in 6.1 https://github.com/rails/rails/commit/7cc27d749c3563e6b278ad01d233cb92ea3b7935 which results in more errors

Can we make MigrationContext a public interface or expose some other "blessed" way to explicitly migrate in tests?

### PR DESCRIPTION
**Pull request by eileencodes**
This change makes MigrationContext public to fix #36544.

* Add docs to MigrationContext and `migrate` method.
* Add `nodoc` to all the private methods for this class.
* Delete filtering behavior from `up`/`down`. This shouldn't be used by
anyone since this was private. It's also an odd feature.
* Delete `run` - all the callers can use `up` or `down` instead of
passing that to `run`.

cc/ @schneems - I finally got around to this 😸

### CODE REVIEW

**Comment by jaredbeck** [comment]
Thanks for this. We've been using `MigrationContext` in the PaperTrail test suite for a while. Prior to 5.2, we used `Migrator`. We use `MigrationContext#migrate`, primarily to test a generator that writes a migration. We test this generator end-to-end, writing the migration to disk and migrating a real db. Not sure if we're doing (any part of) it right, but it's what we do, so (for that dubious reason alone) glad to see it become public.

**Comment by eileencodes** [comment]
I need to table and think about this for a bit. I don't want to make the entire MigrationContext class public because I want to get rid of a bunch of methods in it. In my efforts to reduce its surface area this turned into more of a rabbit hole than I expected. 

Part of me wants to make a new class just for public use that makes a new MigrationContext and only exposes `migrate`, that way I can keep `up`, `down`, `run`, etc private. But it feels wrong to make a new class just for that. `Migrator` and `MigrationContext` seem similar enough they shouldn't be separate but untangling those may take a bit. This wasn't going to make it into 6.1 anyway (already released) so we have time to figure this one out. If you use any other methods besides `new` and `migrate` in `MigrationContext` please let me know here.

**Comment by jaredbeck** [comment]
>> We use MigrationContext#migrate, primarily to test a generator that writes a migration. We test this generator end-to-end, writing the migration to disk and migrating a real db.

> If you use any other methods besides .. migrate .. please let me know ..

We only use `#migrate`, thanks. We test our migration-generator a few times, relying on transaction rollback to clean up our db between tests. Perhaps this desire to use rollback is why we began to use `MigrationContext` programmatically, instead of running `rake db:migrate` in a subshell.

**Review by rafaelfranca — APPROVED** [review]

File: activerecord/lib/active_record/migration.rb [hunk 1]
```diff
@@ -1047,14 +1047,34 @@ def load_migration
...
 
-    def initialize(migrations_paths, schema_migration)
+    def initialize(migrations_paths, schema_migration = SchemaMigration)
       @migrations_paths = migrations_paths
       @schema_migration = schema_migration
     end
 
+    # Runs the migrations in the `migrations_path`.
+    #
+    # If `target_version` is `nil`, `migrate will run `up`.
```
Comment (abhaynikam) [inline activerecord/lib/active_record/migration.rb:1067]: @eileencodes Feels like migrate is missing closing quote(`) here. 😊

Comment (kamipo) [reply activerecord/lib/active_record/migration.rb:1067]: Also use `+` instead of ``` ` ``` in the rdoc.

**Pull request merged**